README

 app_softmodem.c - V.21/V.22/V.23 modem for Asterisk
===============================================================================

	This Asterisk application provides a softmodem for Asterisk. It supports
	the FSK standards V.23, V.21 and Bell 103, but also V.22 and V.22bis.
	Byte-size, number of stopbits and bitordering are configurable.
	
	Two ways are possible to connect to the backend code:
	- A TCP connection comparable to a Telnet session is established to the
	host specified.
	- An executable is forked on the pbx, and communicates via STDIN/STDOUT.


 Dependencies and Installation
-------------------------------------------------------------------------------

	app_softmodem.c depends on the spandsp library
	(http://www.soft-switch.org/).

	It is tested and working with Asterisk 13t.

	To compile put app_softmodem.c in the apps/ dir of your Asterisk source.
	Compile with 'make apps'.

	To use do a 'make install' on your asterisk source or copy
	apps/app_softmodem.so into your asterisk module directory
	(/lib/asterisk/mosules).


 Dialplan
-------------------------------------------------------------------------------

	Application 'Softmodem' with TCP backend
		exten => btx,1,Answer()
		exten => btx,n,Softmodem(host, port, options)
		exten => btx,n,Hangup()

	Application 'SoftmodemExec' with forked binary backend
		exten => btx,1,Answer()
		exten => btx,n,Softmodem(executable, options)
		exten => btx,n,Hangup()

	Without any arguments the application acts as a V.23 modem and connects to a
	Telnet server (port 23) on localhost.

	Options are:
		r(...): rx cutoff (dBi, float, default: -35)
		t(...): tx power (dBi, float, default: -28)
		v(...): modem version (default: V23):
		   	V21        - 300/300 baud
    			V23        - 1200/75 baud
			Bell103    - 300/300 baud
			V22        - 1200/1200 baud
			V22bis     - 2400/2400 baud
		l or m: least or most significant bit first (default: m)
		d(...): amount of data bits (5-8, default: 8)
		s(...): amount of stop bits (1-2, default: 1)
		e or o: add even or odd parity bit (default, none)
		u:      Send ULM header to TCP socket server (Btx, default: false)
		g:      Send generic headers to application (TCP socket or executable, default: false)
		n:      Send NULL-Byte to modem after carrier detection (Btx, default: false)

	To act as a "Btx Vermittlungsstelle" (German videotex system) using V.23, 8 data
	bits, 1 stop bit and sending a NULL byte upon carrier detection and connecting to
	a "twistedUlm" server on port 8289 (thus sending the ULM header):

	exten => 190,1,Softmodem(localhost, 8289, v(V23)ld(8)s(1)un)

	To accept calls from UK Prestel terminals, using V.23, 7 data bits, Even parity,
	1 stop bit:

	exten => 618,n,Softmodem(localhost, 6502, v(V23)ld(7)es(1)t(-8))

	To accept calls from Minitel terminals, using V.23, 7 data bits, Even parity, 1
	stop bit, and Executable backend:

	exten => 3611,n,SoftmodemExec("/srv/minitel/annuaire.php", v(V23)ld(7)es(1)t(-8)g)

	The modem seems to work fine with VOIP as long as you use a codec like G.711
	(alaw/ulaw). Please deactivate any echo cancellation you might use.



 Btx, the Ulm Relay Protocol, general description of operation
-------------------------------------------------------------------------------

	After the softmodem picks up the line a TCP connection to the specified host
	and port is established, the call is dropped when this wasn't successfull.

	When the "n" option is specified a NULL byte is send to the caller upon the
	detection of a carrier signal from the caller. The Btx system in Germany did
	this to request the caller's identification.


news.txt
1 Créteil: taux de radiation habitable

CRETEIL-Le gouvernement a annoncé auj-
ourd'hui que les radiations mesurées
dans les ruines du centre-ville sont
en dessous du seuil de danger pour l'
humain. Le nettoyage des déchets nucl-
éaires causés par l'explosion de la
centrale en Avril 1982.

Le président V. Giscard d'Estaing III
a déclaré: "les efforts de nos scient-
ifiques et de la GISCORP ouvrent la
porte à un réaménagement du sud-est
Parisien, à l'image de La Défense."

2 La recette de la soupe aux truffes VGE

TF1 vous invite à regarder cette recette
en direct samedi soir à 20h. Nous vous
rappelons que la soupe aux truffes était
le plat préféré du président VGE I.

Ingrédients (4 français):
 80g de truffes fraiches
 50cl de bouillon de volaille
 150g de blanc de poulet, sans peau
 100g de céléri-rave
 1 carotte
 8 têtes de champignons de LA DEFENSE
 4 c.à.s de Noilly Prat blanc
 60g de foie gras cuit
 250g de pâte feuilletée
 1 jaune d'œuf
 sel fin

TF1 rappelle que les champignons de LA
DEFENSE sont sans danger d'après le
Ministère de la Santé.

3 Manifestations illégales prévenues

LA DEFENSE-Suite à un rassemblement il-
légal de personnes sur l'ESPLANNADE, la
police a ouvert le feu sur les manifes-
tants. Le groupuscule LA RESISTANCE
avait organisé cette rencontre pour
protester contre la politique du Prési-
dent Valéry Giscard-d'Estaing III.

La fusillade a fait 8 morts et 134
blessés, fort heureusement tous des
manifestants terroristes. Aucun agent
des forces de l'ordre n'a été blessé,
grâce à leurs tactiques et leur nouvel
équipement cybernétique. Plus de 10 000
policiers en cyberarmure avaient été
déployés pour l'occasion, contre 142
manifestants terroristes.

Le droit à manifester a été révoqué
sous la loi martiale de 1983 par le Pr-
ésident VGE III. Les citoyens Français

app_softmodem.c
/*
 * Softmodem for Asterisk
 *
 * 2010, Christian Groeger <code@proquari.at>
 *
 * Based on app_fax.c by Dmitry Andrianov <asterisk@dima.spb.ru>
 * and Steve Underwood <steveu@coppice.org>
 *
 * Parity options added 2018 Rob O'Donnell
 *
 * This program is free software, distributed under the terms of
 * the GNU General Public License
 *
 */

/*** MODULEINFO
	 <depend>spandsp</depend>
	<support_level>extended</support_level>
***/

/* Needed for spandsp headers */
#define ASTMM_LIBC ASTMM_IGNORE

#include "asterisk.h"

//ASTERISK_FILE_VERSION(__FILE__, "$Revision$")

#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <inttypes.h>
#include <pthread.h>
#include <errno.h>
#include <tiffio.h>
#include <time.h>

#include <spandsp.h>
#ifdef HAVE_SPANDSP_EXPOSE_H
#include <spandsp/expose.h>
#endif
#include <spandsp/v22bis.h>

#include "asterisk/file.h"
#include "asterisk/module.h"
#include "asterisk/channel.h"
#include "asterisk/strings.h"
#include "asterisk/lock.h"
#include "asterisk/app.h"
#include "asterisk/pbx.h"
#include "asterisk/format_cache.h"
#include "asterisk/logger.h"
#include "asterisk/utils.h"
#include "asterisk/dsp.h"
#include "asterisk/manager.h"

#ifndef AST_MODULE
#define AST_MODULE "app_softmodem"
#endif

/*** DOCUMENTATION
	<application name="Softmodem" language="en_US">
		<synopsis>
			A Softmodem that connects the caller to a Telnet server.
		</synopsis>
		<syntax />
		<description>
			<para>Softmodem(hostname,port,options):  Simulates a FSK(V.23) or V.22bis modem. The modem on the other end is connected to the specified server using a simple TCP connection (like Telnet).</para>
		</description>
	</application>
 ***/

/**
 * 	"Options:"
	"  r(...): rx cutoff (dBi, float, default: -35)\n"
	"  t(...): tx power (dBi, float, default: -28)\n"
	"  v(...): modem version (default: V22bis):\n"
	"            V21        - 300/300 baud\n"
	"            V23        - 1200/75 baud\n"
	"            Bell103    - 300/300 baud\n"
	"            V22        - 1200/1200 baud\n"
	"            V22bis     - 2400/2400 baud\n"
	"  l or m: least or most significant bit first (default: l)\n"
	"  d(...): amount of data bits (5-8, default: 8)\n"
	"  s(...): amount of stop bits (1-2, default: 1)\n"
	"  e or o: add even or odd parity bit
	"  u:      Send ULM header to Telnet server (Btx)\n"
	"  n:      Send NULL-Byte to modem after carrier detection (Btx)\n" 
	"  c:      This is the calling modem (default)\n" 
	"  a:      This is the answering modem\n" **/

static const char app[] = "Softmodem";

enum {
	OPT_RX_CUTOFF =      (1 << 0),
	OPT_TX_POWER =       (1 << 1),
	OPT_MODEM_VERSION =  (1 << 2),
	OPT_LSB_FIRST =      (1 << 3),
	OPT_MSB_FIRST =      (1 << 4),
	OPT_DATABITS =       (1 << 5),
	OPT_STOPBITS =       (1 << 6),
	OPT_ULM_HEADER =     (1 << 7),
	OPT_NULL =           (1 << 8),
	OPT_CALLER =         (1 << 9),
	OPT_ANSWERER =       (1 << 10)
	OPT_EVEN_PARITY =    (1 << 11),
	OPT_ODD_PARITY =     (1 << 12)
};

enum {
	OPT_ARG_RX_CUTOFF,
	OPT_ARG_TX_POWER,
	OPT_ARG_MODEM_VERSION,
	OPT_ARG_DATABITS,
	OPT_ARG_STOPBITS,

	/* Must be the last element */
	OPT_ARG_ARRAY_SIZE,
};

AST_APP_OPTIONS(additional_options, BEGIN_OPTIONS
	AST_APP_OPTION_ARG('r', OPT_RX_CUTOFF, OPT_ARG_RX_CUTOFF),
	AST_APP_OPTION_ARG('t', OPT_TX_POWER, OPT_ARG_TX_POWER),
	AST_APP_OPTION_ARG('v', OPT_MODEM_VERSION, OPT_ARG_MODEM_VERSION),
	AST_APP_OPTION('l', OPT_LSB_FIRST),
	AST_APP_OPTION('m', OPT_MSB_FIRST),
	AST_APP_OPTION_ARG('d', OPT_DATABITS, OPT_ARG_DATABITS),
	AST_APP_OPTION_ARG('s', OPT_STOPBITS, OPT_ARG_STOPBITS),
	AST_APP_OPTION('e', OPT_EVEN_PARITY),
	AST_APP_OPTION('o', OPT_ODD_PARITY),
	AST_APP_OPTION('u', OPT_ULM_HEADER),
	AST_APP_OPTION('n', OPT_NULL),
	AST_APP_OPTION('c', OPT_CALLER),
	AST_APP_OPTION('a', OPT_ANSWERER),
END_OPTIONS );


#define MAX_SAMPLES 240

enum {
	VERSION_V21,
	VERSION_V23,
	VERSION_BELL103,
	VERSION_V22,
	VERSION_V22BIS,
};

typedef struct {
	struct ast_channel *chan;
	const char *host;	//telnetd host+port
	int port;
	float txpower;
	float rxcutoff;
	int version;
	int lsb;
	int databits;
	int stopbits;
	int ulmheader;
	int sendnull;
	int caller;
	volatile int finished;
	int	paritytype;
} modem_session;

#define MODEM_BITBUFFER_SIZE 16
typedef struct {
	int answertone;		// terminal is active (=sends data)
	int nulsent;		// we sent a NULL as very first character (at least the DBT03 expects this)
} connection_state;
typedef struct {
	int sock;
	int bitbuffer[MODEM_BITBUFFER_SIZE];
	int writepos;
	int readpos;
	int fill;
	connection_state *state;
	modem_session *session;
} modem_data;


// this is called by spandsp whenever it filtered a new bit from the line
static void modem_put_bit(void *user_data, int bit) {
	int stop, stop2, i;

	modem_data *rx = (modem_data*) user_data;

	int databits = rx->session->databits;
	int stopbits = rx->session->stopbits;
	int paritybits = 0;

	if (rx->session->paritytype)
		paritybits = 1;

	// modem recognised us and starts responding through sending it's pilot signal
	if (rx->state->answertone<=0) {
		if (bit==SIG_STATUS_CARRIER_UP) { 
			ast_debug(1, "Carrier up, setting answertone = 0\n");
			rx->state->answertone=0;
		}
		if (rx->session->version == VERSION_V22 || rx->session->version == VERSION_V22BIS) {
			if (bit==SIG_STATUS_TRAINING_SUCCEEDED && rx->state->answertone==0) {
				ast_debug(1, "V22 training succeeded, setting answertone = 1\n");
				rx->state->answertone=1;
				if (rx->session->caller) send(rx->sock, "CONNECTED\r\n", 11, 0);
			}
		} else {
			if (bit==1 && rx->state->answertone==0) {
				ast_debug(1, "Non-V22 started to receive data, setting answertone = 1\n");
				rx->state->answertone=1;
				if (rx->session->caller) send(rx->sock, "CONNECTED\r\n", 11, 0);
			}
		}
	}
	else {
		// ignore other spandsp-stuff
		if (bit==1 || bit==0) {

			// insert bit into our bitbuffer
			rx->bitbuffer[rx->writepos]=bit;
			rx->writepos++;
			if (rx->writepos>=MODEM_BITBUFFER_SIZE) rx->writepos=0;

			if (rx->fill<MODEM_BITBUFFER_SIZE) {
				rx->fill++;
			} else {
				// our bitbuffer is full, this probably won't happen
	// 			printf("full buffer\n");
				rx->readpos++;
				if (rx->readpos>=MODEM_BITBUFFER_SIZE) rx->readpos=0;
			}

			// full byte = 1 startbit + databits + paritybits + stopbits
			while (rx->fill>=(1+databits+paritybits+stopbits)) {
				if (rx->bitbuffer[rx->readpos]==0) {	// check for startbit
					stop=(rx->readpos+1+paritybits+databits)%MODEM_BITBUFFER_SIZE;
					stop2=(rx->readpos+2+paritybits+databits)%MODEM_BITBUFFER_SIZE;
					if ( (rx->bitbuffer[stop]==1) &&
						(stopbits==1 || (stopbits==2 && rx->bitbuffer[stop2]==1)) )
					{	// check for stopbit -> valid framing


						char byte=0;

						for(i=0; i<databits; i++) {	// generate byte
							if (rx->session->lsb) { //lsb first
								if (rx->bitbuffer[(rx->readpos+1+i)%MODEM_BITBUFFER_SIZE])
									byte |= (1<<i);
							} else { //msb first
								if (rx->bitbuffer[(rx->readpos+databits-i)%MODEM_BITBUFFER_SIZE])
									byte |= (1<<i);
							}
						}

						if ( !paritybits || ( paritybits &&
									( rx->bitbuffer[(rx->readpos + databits + 1)%MODEM_BITBUFFER_SIZE] ==
										( (rx->session->paritytype == 2) ^ __builtin_parity(byte) ) ) ) ) {

							send(rx->sock, &byte, 1, 0);
						} // else invalid parity, ignore byte
						// TODO - why does this increment by 10?
						rx->readpos=(rx->readpos+10)%MODEM_BITBUFFER_SIZE;
						rx->fill-=10;
					} else {	// no valid framing (no stopbit), remove first bit and maybe try again
						rx->fill--;
						rx->readpos++;
						rx->readpos %= MODEM_BITBUFFER_SIZE;
					}
				} else {	// no valid framing either (no startbit)
					rx->fill--;
					rx->readpos++;
					rx->readpos %= MODEM_BITBUFFER_SIZE;
				}
			}
		}
		else {
			ast_debug(1, "Received a non 1 or 0 bit: %d\n", bit);
		}
	}



	return;
}

// spandsp asks us for a bit to send onto the line
static int modem_get_bit(void *user_data) {
	modem_data *tx = (modem_data*) user_data;
	char byte=0;
	int i, rc;

	int databits=tx->session->databits;
	int stopbits=tx->session->stopbits;
	int paritybits = 0;

	if (tx->session->paritytype)
		paritybits = 1;



	// no new data in send (bit)buffer,
	// either we just picked up the line, the terminal started to respond,
	// than we check for new data on the socket
	// or there's no new data, so we send 1s (mark)
	if (tx->writepos==tx->readpos) {
		if(tx->state->nulsent>0) {	// connection is established, look for data on socket
			rc=recv(tx->sock,&byte, 1, 0);
			if (rc>0) {
				// new data on socket, we put that byte into our bitbuffer
				for (i=0; i<(databits+paritybits+stopbits); i++) {
					if (paritybits && (i == databits) ) {
						tx->bitbuffer[tx->writepos] = (tx->session->paritytype == 2) ^ __builtin_parity( byte);
					} else if ( i >= databits )
						tx->bitbuffer[tx->writepos]=1;	// stopbits
					else {	// databits
						if (tx->session->lsb) {
							if (byte & (1<<i)) tx->bitbuffer[tx->writepos]=1;
							else tx->bitbuffer[tx->writepos]=0;
						} else {
							if (byte & (1<<(databits-1-i))) tx->bitbuffer[tx->writepos]=1;
							else tx->bitbuffer[tx->writepos]=0;
						}
					}
					tx->writepos++;
					if (tx->writepos>=MODEM_BITBUFFER_SIZE) tx->writepos=0;
				}
				return 0; // return startbit immediately
			}
			else if (rc==0) {
				ast_log(LOG_WARNING,"Socket seems closed. Will hangup.\n");
				tx->session->finished=1;
			}
		}
		else {
			// check if socket was closed before connection was terminated
			if ( recv(tx->sock,&byte, 1, MSG_PEEK) == 0 ) {
				tx->session->finished=1;
				return 1;
			}
			if ( tx->state->answertone>0 ) {
				if (tx->session->sendnull) { // send null byte
					for (i=0; i<(databits+paritybits+stopbits); i++) {
						if (paritybits && (i == databits) )
							tx->bitbuffer[tx->writepos] = (tx->session->paritytype == 2) ^ __builtin_parity( 0 );	// yes I know!
						else if (i>=databits) tx->bitbuffer[tx->writepos]=1;	//stopbits
						else tx->bitbuffer[tx->writepos]=0; //databits
						tx->writepos++;
						if (tx->writepos>=MODEM_BITBUFFER_SIZE) tx->writepos=0;
					}
				}
				tx->state->nulsent=1;

				if (tx->session->ulmheader) {
					// send ULM relay protocol header, include connection speed
					float tx_baud,rx_baud;
					if (tx->session->version==VERSION_V21) {
						tx_baud=300;
						rx_baud=300;
					} else if (tx->session->version==VERSION_V23) {
						tx_baud=1200;
						rx_baud=75;
					} else if (tx->session->version==VERSION_BELL103) {
						tx_baud=300;
						rx_baud=300;
					} else if (tx->session->version==VERSION_V22) {
						tx_baud=1200;
						rx_baud=1200;
					} else if (tx->session->version==VERSION_V22BIS) {
						tx_baud=2400;
						rx_baud=2400;
					} else {
						tx_baud=0;
						rx_baud=0;
					}

					char header[60];
					int headerlength=sprintf(header,
						"Version: 1\r\nTXspeed: %.2f\r\nRXspeed: %.2f\r\n\r\n",
						tx_baud/(1+databits+stopbits), rx_baud/(1+databits+stopbits));
					send(tx->sock, header, headerlength, 0);
				}

				if (tx->session->sendnull)
					return 0;
				else
					return 1;
			}
		}

		// no new data on socket, NULL-byte already sent, send mark-frequency
		return 1;

	} else {
		// there still is data in the bitbuffer, so we just send that out
		i=tx->bitbuffer[tx->readpos];
		tx->readpos++;
		if (tx->readpos>=MODEM_BITBUFFER_SIZE) tx->readpos=0;
// 		printf("b%i",i);
		return i;
	}
}

static void *modem_generator_alloc(struct ast_channel *chan, void *params)
{
	return params;
}

static int fsk_generator_generate(struct ast_channel *chan, void *data, int len, int samples)
{
	fsk_tx_state_t *tx = (fsk_tx_state_t*) data;
	uint8_t buffer[AST_FRIENDLY_OFFSET + MAX_SAMPLES * sizeof(uint16_t)];
	int16_t *buf = (int16_t *) (buffer + AST_FRIENDLY_OFFSET);

	struct ast_frame outf = {
		.frametype = AST_FRAME_VOICE,
		.subclass.format = ast_format_slin,
		.src = __FUNCTION__,
	};

	if (samples > MAX_SAMPLES) {
		ast_log(LOG_WARNING, "Only generating %d samples, where %d requested\n", MAX_SAMPLES, samples);
		samples = MAX_SAMPLES;
	}


	if ((len = fsk_tx(tx, buf, samples)) > 0) {
		outf.samples = len;
		AST_FRAME_SET_BUFFER(&outf, buffer, AST_FRIENDLY_OFFSET, len * sizeof(int16_t));

		if (ast_write(chan, &outf) < 0) {
			ast_log(LOG_WARNING, "Failed to write frame to %s: %s\n", ast_channel_name(chan), strerror(errno));
			return -1;
		}
	}

	return 0;
}


static int v22_generator_generate(struct ast_channel *chan, void *data, int len, int samples)
{
	v22bis_state_t *tx = (v22bis_state_t*) data;
	uint8_t buffer[AST_FRIENDLY_OFFSET + MAX_SAMPLES * sizeof(uint16_t)];
	int16_t *buf = (int16_t *) (buffer + AST_FRIENDLY_OFFSET);

	struct ast_frame outf = {
		.frametype = AST_FRAME_VOICE,
		.subclass.format = ast_format_slin,
		.src = __FUNCTION__,
	};

	if (samples > MAX_SAMPLES) {
		ast_log(LOG_WARNING, "Only generating %d samples, where %d requested\n", MAX_SAMPLES, samples);
		samples = MAX_SAMPLES;
	}


	if ((len = v22bis_tx(tx, buf, samples)) > 0) {
		outf.samples = len;
		AST_FRAME_SET_BUFFER(&outf, buffer, AST_FRIENDLY_OFFSET, len * sizeof(int16_t));

		if (ast_write(chan, &outf) < 0) {
			ast_log(LOG_WARNING, "Failed to write frame to %s: %s\n", ast_channel_name(chan), strerror(errno));
			return -1;
		}
	}

	return 0;
}

struct ast_generator fsk_generator = {
	alloc:		modem_generator_alloc,
	generate: 	fsk_generator_generate,
};
struct ast_generator v22_generator = {
	alloc:		modem_generator_alloc,
	generate: 	v22_generator_generate,
};

static int softmodem_communicate(modem_session *s) {
	int res = -1;
	struct ast_format *original_read_fmt;
	struct ast_format *original_write_fmt;

	modem_data rxdata, txdata;

	struct ast_frame *inf = NULL;

	fsk_tx_state_t *modem_tx;
	fsk_rx_state_t *modem_rx;

	v22bis_state_t *v22_modem;


	original_read_fmt = ast_channel_readformat(s->chan);
	if (original_read_fmt != ast_format_slin) {
		res=ast_set_read_format(s->chan, ast_format_slin);
		if (res < 0) {
			ast_log(LOG_WARNING, "Unable to set to linear read mode, giving up\n");
			return res;
		}
	}

	original_write_fmt = ast_channel_writeformat(s->chan);
	if (original_write_fmt != ast_format_slin) {
		res=ast_set_write_format(s->chan, ast_format_slin);
		if (res < 0) {
			ast_log(LOG_WARNING, "Unable to set to linear write mode, giving up\n");
			return res;
		}
	}

	int sock;
	struct sockaddr_in server;
	struct hostent *hp;
	struct ast_hostent ahp;

	sock = socket(AF_INET, SOCK_STREAM, 0);
	if(sock < 0) {
		ast_log(LOG_WARNING, "Could not create socket.\n");
		return res;
	}

	server.sin_family=AF_INET;
	hp=ast_gethostbyname(s->host, &ahp);
	memcpy( (char *)&server.sin_addr, hp->h_addr, hp->h_length);
	//bcopy ( hp->h_addr, &(server.sin_addr.s_addr), hp->h_length);
	server.sin_port=htons(s->port);

	if (connect(sock, (struct sockaddr*)&server, sizeof(server)) < 0) {
		ast_log(LOG_WARNING, "Cannot connect to remote host.\n");
		return res;
	}

	fcntl(sock, F_SETFL, O_NONBLOCK);


	connection_state state;
	state.answertone=-1;	//no carrier yet
	state.nulsent=0;

	rxdata.sock=sock;
	rxdata.writepos=0;
	rxdata.readpos=0;
	rxdata.fill=0;
	rxdata.state= &state;
	rxdata.session=s;

	txdata.sock=sock;
	txdata.writepos=0;
	txdata.readpos=0;
	txdata.fill=0;
	txdata.state= &state;
	txdata.session=s;

	// initialise spandsp-stuff, give it our callback-functions
	if (s->version==VERSION_V21) {
		ast_debug(1, "Initializing V21\n");
		if (s->caller) {
			modem_tx = fsk_tx_init(NULL, &preset_fsk_specs[FSK_V21CH2], modem_get_bit, &txdata);
			modem_rx = fsk_rx_init(NULL, &preset_fsk_specs[FSK_V21CH1], TRUE, modem_put_bit, &rxdata);
		} else {
			modem_tx = fsk_tx_init(NULL, &preset_fsk_specs[FSK_V21CH1], modem_get_bit, &txdata);
			modem_rx = fsk_rx_init(NULL, &preset_fsk_specs[FSK_V21CH2], TRUE, modem_put_bit, &rxdata);
		}
	}
	else if (s->version==VERSION_V23) {
		ast_debug(1, "Initializing V23\n");
		if (s->caller) {
			modem_tx = fsk_tx_init(NULL, &preset_fsk_specs[FSK_V23CH2], modem_get_bit, &txdata);
			modem_rx = fsk_rx_init(NULL, &preset_fsk_specs[FSK_V23CH1], TRUE, modem_put_bit, &rxdata);
		} else {
			modem_tx = fsk_tx_init(NULL, &preset_fsk_specs[FSK_V23CH1], modem_get_bit, &txdata);
			modem_rx = fsk_rx_init(NULL, &preset_fsk_specs[FSK_V23CH2], TRUE, modem_put_bit, &rxdata);
		}	
	}
	else if (s->version==VERSION_BELL103) {
		ast_debug(1, "Initializing Bell103\n");
		if (s->caller) {
			modem_tx = fsk_tx_init(NULL, &preset_fsk_specs[FSK_BELL103CH2], modem_get_bit, &txdata);
			modem_rx = fsk_rx_init(NULL, &preset_fsk_specs[FSK_BELL103CH1], FALSE, modem_put_bit, &rxdata);
		} else {
			modem_tx = fsk_tx_init(NULL, &preset_fsk_specs[FSK_BELL103CH1], modem_get_bit, &txdata);
			modem_rx = fsk_rx_init(NULL, &preset_fsk_specs[FSK_BELL103CH2], FALSE, modem_put_bit, &rxdata);
		}
	}
	else if (s->version==VERSION_V22) {
		ast_debug(1, "Initializing V22\n");
		v22_modem = v22bis_init(NULL, 1200, 0, s->caller, modem_get_bit, &txdata, modem_put_bit, &rxdata);
	}
	else if (s->version==VERSION_V22BIS) {
		ast_debug(1, "Initializing V22bis\n");
		v22_modem = v22bis_init(NULL, 2400, 0, s->caller, modem_get_bit, &txdata, modem_put_bit, &rxdata);
	}
	else {
		ast_log(LOG_ERROR,"Unsupported modem type. Sorry.\n");
		return res;
	}

	if (s->version==VERSION_V21 || s->version==VERSION_V23 || s->version==VERSION_BELL103) {
		fsk_tx_power (modem_tx, s->txpower);
		fsk_rx_signal_cutoff(modem_rx, s->rxcutoff);
	}
	else if (s->version==VERSION_V22 || s->version==VERSION_V22BIS) {
		v22bis_tx_power(v22_modem, s->txpower);
		v22bis_rx_signal_cutoff(v22_modem, s->rxcutoff);
	}

	//printf("comm: baud %i\n",btx_tx->baud_rate);
	if (s->version==VERSION_V21 || s->version==VERSION_V23 || s->version==VERSION_BELL103)
		ast_activate_generator(s->chan, &fsk_generator, modem_tx);
	else if (s->version==VERSION_V22 || s->version==VERSION_V22BIS)
		ast_activate_generator(s->chan, &v22_generator, v22_modem);

	while (!s->finished) {
		res = ast_waitfor(s->chan, 20);
		if (res < 0)
			break;
		else if (res > 0)
			res = 0;

		inf = ast_read(s->chan);
		if (inf == NULL) {
			ast_debug(1, "Channel hangup\n");
			res = -1;
			break;
		}

		/* Check the frame type. Format also must be checked because there is a chance
		   that a frame in old format was already queued before we set chanel format
		   to slinear so it will still be received by ast_read */
		if (inf->frametype == AST_FRAME_VOICE && inf->subclass.format == ast_format_slin) {
			if (s->version==VERSION_V21 || s->version==VERSION_V23 || s->version==VERSION_BELL103) {
				if (fsk_rx(modem_rx, inf->data.ptr, inf->samples) < 0) {
					/* I know fax_rx never returns errors. The check here is for good style only */
					ast_log(LOG_WARNING, "softmodem returned error\n");
					res = -1;
					break;
				}
			}
			else if (s->version==VERSION_V22 || s->version==VERSION_V22BIS) {
				if (v22bis_rx(v22_modem, inf->data.ptr, inf->samples) < 0) {
					ast_log(LOG_WARNING, "softmodem returned error\n");
					res = -1;
					break;
				}
			}
		}

		ast_frfree(inf);
		inf = NULL;
	}

	close(sock);

	if (s->version==VERSION_V22 || s->version==VERSION_V22BIS) {
		v22bis_release(v22_modem);
		v22bis_free(v22_modem);
	}

	if (original_write_fmt != ast_format_slin) {
		if (ast_set_write_format(s->chan, original_write_fmt) < 0)
			ast_log(LOG_WARNING, "Unable to restore write format on '%s'\n", ast_channel_name(s->chan));
	}

	if (original_read_fmt != ast_format_slin) {
		if (ast_set_read_format(s->chan, original_read_fmt) < 0)
			ast_log(LOG_WARNING, "Unable to restore read format on '%s'\n", ast_channel_name(s->chan));
	}

	return res;

}

static int softmodem_exec(struct ast_channel *chan, const char *data) {
	int res = 0;
	char *parse;
	modem_session session;
	struct ast_flags options = { 0, };
	char *option_args[OPT_ARG_ARRAY_SIZE];

	AST_DECLARE_APP_ARGS(args,
		AST_APP_ARG(host);
		AST_APP_ARG(port);
		AST_APP_ARG(options);
	);

	if (chan == NULL) {
		ast_log(LOG_ERROR, "Channel is NULL. Giving up.\n");
		return -1;
	}

	/* answer channel if not already answered */
	if (ast_channel_state(chan) != AST_STATE_UP) {
		res = ast_answer(chan);
		if (res) {
			ast_log(LOG_WARNING, "Could not answer channel '%s'\n", ast_channel_name(chan));
			return res;
		}
	}

	session.chan=chan;
	session.finished=0;
	session.rxcutoff=-35.0f;
	session.txpower=-28.0f;
	session.version=VERSION_V22BIS;
	session.lsb=1;
	session.databits=8;
	session.stopbits=1;
	session.paritytype=0;
	session.ulmheader=0;
	session.sendnull=0;
	session.caller=1;

	parse=ast_strdupa(data);
	AST_STANDARD_APP_ARGS(args,parse);

	if (args.host)
		session.host=args.host;
	else
		session.host="localhost";

	if (args.port) {
		session.port=atoi(args.port);
		if ((session.port<0) || (session.port>65535)) {
			ast_log(LOG_ERROR, "Please specify a valid port number.\n");
			return -1;
		}
	} else
		session.port=23;


	if (args.options) {
		ast_app_parse_options(additional_options, &options, option_args, args.options);

		if (ast_test_flag(&options, OPT_RX_CUTOFF)) {
			if (!ast_strlen_zero(option_args[OPT_ARG_RX_CUTOFF]))
				session.rxcutoff=atof(option_args[OPT_ARG_RX_CUTOFF]);
		}

		if (ast_test_flag(&options, OPT_TX_POWER)) {
			if (!ast_strlen_zero(option_args[OPT_ARG_TX_POWER]))
				session.txpower=atof(option_args[OPT_ARG_TX_POWER]);
		}

		if (ast_test_flag(&options, OPT_MODEM_VERSION)) {
			if (!ast_strlen_zero(option_args[OPT_ARG_MODEM_VERSION])) {
				ast_debug(1, "Modem version from args is %s\n", option_args[OPT_ARG_MODEM_VERSION]);
				if (strcmp(option_args[OPT_ARG_MODEM_VERSION],"V21")==0)
					session.version=VERSION_V21;
				else if (strcmp(option_args[OPT_ARG_MODEM_VERSION],"V23")==0)
					session.version=VERSION_V23;
				else if (strcmp(option_args[OPT_ARG_MODEM_VERSION],"Bell103")==0)
					session.version=VERSION_BELL103;
				else if (strcmp(option_args[OPT_ARG_MODEM_VERSION],"V22")==0)
					session.version=VERSION_V22;
				else if (strcmp(option_args[OPT_ARG_MODEM_VERSION],"V22bis")==0)
					session.version=VERSION_V22BIS;
			}
		}

		if (ast_test_flag(&options, OPT_LSB_FIRST)) {
			if (ast_test_flag(&options, OPT_MSB_FIRST)) {
				ast_log(LOG_ERROR, "Please only set l or m flag, not both.\n");
				return -1;
			}
			session.lsb=1;
		}

		if (ast_test_flag(&options, OPT_DATABITS)) {
			if (!ast_strlen_zero(option_args[OPT_ARG_DATABITS])) {
				session.databits = atoi(option_args[OPT_ARG_DATABITS]);

				if ((session.databits<5) || (session.databits>8)) {
					ast_log(LOG_ERROR, "Only 5-8 data bits are supported.\n");
					return -1;
				}
			}
		}

		if (ast_test_flag(&options, OPT_STOPBITS)) {
			if (!ast_strlen_zero(option_args[OPT_ARG_STOPBITS])) {
				session.stopbits = atoi(option_args[OPT_ARG_STOPBITS]);
				if ((session.stopbits<1) || (session.stopbits>2)) {
					ast_log(LOG_ERROR, "Only 1-2 stop bits are supported.\n");
					return -1;
				}
			}
		}

		if (ast_test_flag(&options, OPT_EVEN_PARITY)) {
			if (ast_test_flag(&options, OPT_ODD_PARITY)) {
				ast_log(LOG_ERROR, "Please only set e or o (parity) flag, not both.\n");
				return -1;
			}
			session.paritytype = 1;
		}
		if (ast_test_flag(&options, OPT_ODD_PARITY)) {
			session.paritytype = 2;
		}

		if (ast_test_flag(&options, OPT_ULM_HEADER)) {
			session.ulmheader = 1;
		}
		if (ast_test_flag(&options, OPT_NULL)) {
			session.sendnull = 1;
		}
		if (ast_test_flag(&options, OPT_ANSWERER)) {
			if (ast_test_flag(&options, OPT_CALLER)) {
				ast_log(LOG_ERROR, "Please only set c or a flag, not both.\n");
				return -1;
			}
			session.caller = 0;
		}
		
	}

	res=softmodem_communicate(&session);

	return res;
}



static int unload_module(void) {
	return ast_unregister_application(app);
}

static int load_module(void) {
	return ast_register_application_xml(app, softmodem_exec);
}


AST_MODULE_INFO_STANDARD_EXTENDED(ASTERISK_GPL_KEY, "Softmodem v22");
